### PR TITLE
Replace tmpdir with tmp_path

### LINKED ISSUE #9662: Use `tmp_path` instead of `tmpdir` pytest fixture
Just a heads up that the `tmpdir` fixture has been replaced by `tmp_path` (see [here](https://docs.pytest.org/en/6.2.x/fixture.html#:~:text=pathlib.Path%20objects.-,tmpdir,-Provide%20a%20py)). New tests should use `tmp_path`, and if someone has the time to search and replace, we could also completely switch to the recommended fixture.

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes #9662.

### CODE REVIEW

**Comment by cbrnr** [comment]
This has not been super fun (simply replacing `tmpdir` with `tmp_path` didn't cut it), but now it is almost working except for the following two issues (@larsoner maybe you can help):

Windows is crashing [here](https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=16350&view=logs&j=fa98c7b3-575e-55b6-6946-22dea6f6daf8&t=9997f553-f8cd-520b-0850-4d0fe0dea652&l=225) ("Windows fatal exception: access violation"), any ideas?

There's one failing test that I haven't fixed yet, because when saving an already loaded `Raw` object, it now raises a `FileExistsError` in https://github.com/mne-tools/mne-python/blob/main/mne/utils/check.py#L167, but `test_save` expects a different error (https://github.com/cbrnr/mne-python/blob/243da7562012baf4622ae9f8752557eb37709e72/mne/io/fiff/tests/test_raw_fiff.py#L1392) that used to be raised [here](https://github.com/cbrnr/mne-python/blob/fde89c53dedd89338346bebd1ac67335fb80375f/mne/io/base.py#L1428). How would you like me to proceed?

**Comment by larsoner** [comment]
> There's one failing test that I haven't fixed yet, because when saving an already loaded Raw object, it now raises a FileExistsError

This to me suggests that there is some problem with the first check that's supposed to see "if data are not preloaded and we're trying to save to self._filenames[0], raise a ValueError", and instead you hit the second / subsequent check that is "if the filename exists and overwrite is False (default), then raise a FileExistsError". I would look into why using this new fixture has changed the behavior of the first check

File: mne/io/base.py
Comment (larsoner) [inline mne/io/base.py:1427]: Nowadays we don't cast using `str` but rather use `fname = _check_fname(fname)` because it does some nice standardization for us

Comment (cbrnr) [reply mne/io/base.py]: Done.

**Comment by cbrnr** [comment]
@larsoner the issue is that `fname = _check_fname(fname)` raises a `FileExistsError: Destination file exists. Please use option "overwrite=True" to force overwriting.` even before the other exception (`if not self.preload and fname in self._filenames:`) can be raised. I could of course set `overwrite=True`, but I guess this defeats the purpose of the other check. Any suggestions on how to proceed?

**Comment by larsoner** [comment]
> I could of course set overwrite=True, but I guess this defeats the purpose of the other check.

I would use `overwrite=True` because you don't want/need *that* function to do an overwrite check, because you're doing a more targeted one later

**Comment by cbrnr** [comment]
This is now ready for review. Maybe worth a changelog entry too, but we can do this as the final step.

**Comment by agramfort** [comment]
@cbrnr can we now remove tmpdir from our fixtures? Did you already take care of this? 🙏 for the cleanup

**Comment by cbrnr** [comment]
What do you mean by remove? I replaced `tmpdir` with `tmp_path` everywhere, is there some config file I need to update?

**Comment by sappelhoff** [comment]
And if you're in the flow of it @cbrnr perhaps you wanna do a quick pass on mne-bids as well, replacing the tmpdir -> tmp_path there? :smirk: 

But I see your comment:

> This has not been super fun

so if you've had enough for now, we'll just do it some other time when one of us has time and energy

**Comment by agramfort** [comment]
indeed tmpdir is a pytest builtin fixture.

git grep still says:

mne/gui/tests/test_coreg_gui.py:def
test_coreg_gui_pyvista_file_support(inst_path, tmpdir,
mne/gui/tests/test_coreg_gui.py:    tempdir = str(tmpdir)
mne/gui/tests/test_coreg_gui.py:def test_coreg_gui_pyvista(tmpdir,
renderer_interactive_pyvistaqt):
mne/gui/tests/test_coreg_gui.py:    tempdir = str(tmpdir)

is it expected to keep it there?

>

**Comment by cbrnr** [comment]
@agramfort thanks for double-checking, I missed those. Now there are no `tmpdir` fixtures anymore.

@sappelhoff I can take a look, mne-bids is much smaller and therefore this should be much faster.

**Comment by cbrnr** [comment]
@sappelhoff https://github.com/mne-tools/mne-bids/pull/896

**Comment by cbrnr** [comment]
Any idea what's going on with CircleCI now? https://app.circleci.com/pipelines/github/mne-tools/mne-python/11173/workflows/849f4079-3350-4539-be79-8cdff6404ee1/jobs/37508/parallel-runs/0/steps/0-103

**Comment by cbrnr** [comment]
@larsoner now in addition to CircleCI most (all?) Azure jobs are failing because of an issue with OpenGL install?

**Comment by larsoner** [comment]
#9943

**Comment by cbrnr** [comment]
Please ping me once that is merged.

**Comment by cbrnr** [comment]
(This might be something where [merge queues](https://github.blog/changelog/2021-10-27-pull-request-merge-queue-limited-beta/) could be useful – we should definitely try them when they are available.)

**Comment by larsoner** [comment]
> Please ping me once that is merged.

If you go to the PR and click the "subscribe" you'll get notified automatically

**Comment by cbrnr** [comment]
@GuillaumeFavelier do you know why `coreg._lock_fids` is not `False`? https://github.com/mne-tools/mne-python/runs/4081446793?check_suite_focus=true#step:13:3462

File: mne/tests/test_source_estimate.py
Comment (drammock) [inline mne/tests/test_source_estimate.py:440]:
```suggestion
    # different errors for (non)vector STCs
    match = (r"(Invalid value for the 'ftype' par)|"
             r"(VectorSourceEstimate objects can only be written ad HDF5)*")
    with pytest.raises(ValueError, match=match):
        stc.save(tmp_path / 'tmp', ftype='foo')
```

**Review by drammock — APPROVED** [review]

File: mne/tests/test_source_estimate.py
Comment (cbrnr) [reply mne/tests/test_source_estimate.py:440]: Should we go with this more specific error message catching or wait for the fix in #9944?

Comment (larsoner) [reply mne/tests/test_source_estimate.py:440]: I don't think my fix in #9944 will impact this comment -- it looks like this comment just improves/modernizes the `pytest.raises` usage. So yes feel free to commit this suggestion separately

**Review by larsoner — APPROVED** [review]
LGTM, will merge once CIs come back happy. Thanks @cbrnr

**Comment by agramfort** [comment]
thx @cbrnr !

**Pull request merged**
